proposals/lip-update_block_schema_and_block_processing.md

```diff
@@ -0,0 +1,524 @@
+Each module can insert a single entry in the assets.
+This entry is an object containing a `moduleID` property, indicating the ID of the module handling it, and a generic `data` property that can contain arbitrary serialized data.
+
+Each entry of the block assets is then inserted in a Merkle tree, whose root is included in the block header as the `assetsRoot` property.
+Inserting the assets root rather than the full assets allows to bound the size of the block header while still authenticating the content of the block assets.
+
+
+## Specification
+
+### Notation and Constants
+
+For the rest of this proposal we define the following constants.
+
+| Name          | Type    | Value       | Description |
+| ------------- |---------| ------------| ------------|
+| `MAX_PAYLOAD_SIZE_BYTES` | integer | TBD | The max size of a block payload in bytes.|
+| `MAX_ASSET_DATA_SIZE_BYTES` | integer | TBD | The max size of an assets entry in bytes.|
+|`SIGNATURE_LENGTH_BYTES`| integer | 64 | The length of a Ed25519 signature.|
+
+
+Furthermore, in the following we indicate with `block` be the block under consideration and with `previousBlock` the previous block of the chain.
+Calling a function `fct` from another module `module` is represented by `module.fct`.
+
+### Block
+
+#### JSON Schema
+
+Blocks are serialized and deserialized accordingly to the following JSON schema.
+
+```java
+blockSchema = {
+  "type": "object",
+  "required": ["header", "payload", "assets"],
+  "properties": {
+    "header": {
+      "dataType": "bytes",
+      "fieldNumber": 1
+    },
+    "payload": {
+      "type": "array",
+      "fieldNumber": 2,
+      "items": {
+        "dataType": "bytes"
+      }
+    },
+    "assets": {
+      "type": "array",
+      "fieldNumber": 3,
+      "items": {
+        "dataType": "bytes"
+      }
+    }
+  }
+}
+```
+
+#### Validation
+
+Blocks are validated at three different stages of their lifecycle.
+
+- **Static validation**: When a new block is received, some initial static checks are done to ensure that the serialized object follows the general structure of a block. 
+  These checks are performed immediately because they do not require access to the state store and can therefore be done very quickly.
+- **Block verification**: Properties that require access to the state store *before* the block has been executed are verified in this stage.
+In particular, these checks are performed *before* the state transitions implied by the modules are processed.
+- **Result verification**: In this stage we verify the properties that require access to the state store *after* the block has been executed, i.e. they can be verified only after the state transitions implied by the block execution have been performed.
+In particular, these checks are performed *after* the state transitions implied by the modules are processed.
```

Comment:
It seems that lines 143 to 150 concern the validation of the block, block header and block asset. Therefore I would think it makes sense to extract it to a separate level 3 section just after the section "Notation and Constants". The section could be called "Validation stages of block, block asset and block header`.

Reply:
Moved and slightly rephrased

Reply:
Can be resolved.